### ISSUE TITLE
[Keybase] Miniature Keybase Implementation to Enable M1 - Create and Delete

### ISSUE THREAD

**Issue by andrewnguyen22** [pokt-network/pocket#150]
# Objective
Create a miniature Keybase that can create, store, and delete keys

# Origin Document
In order to achieve the M1 milestone a mini-keybase implementation is needed to enable dynamic accounts and actors.

This keybase may be used for the following tasks:
- Signing blocks
- Signing transactions
- Signing arbitrary messages
- Signing multi-sig messages

# Goals / Deliverables
Mini ED-25519 Keybase: 
- [ ] Code complete implementation of:
   - [ ] **CREATE** and store keys
   - [ ] **DELETE** keys
- [ ] Document the key architecture 

# Non-Goals / Deliverables
- Integrating Keybase with the CLI or other parts of the codebase
- Implementing multisig

## General issue checklist
- [ ] Update the appropriate CHANGELOG
- [ ] Update the README
- [ ] If applicable, update the source code tree explanation
- [ ] If applicable, add or update a state, sequence or flowchart diagram using [mermaid](https://mermaid-js.github.io/mermaid/)
- [ ] Update any relevant global documentation & references
- [ ] Document small issues / TODOs along the way

---

Creator: @andrewnguyen22

**Comment by andrewnguyen22** [comment]
**NOTE**: Please use the `/shared/crypto/ed25519` as the P/K library

**Comment by andrewnguyen22** [comment]
**Recommend** to use a filesystem db for the implementation

Currently we are between `Badger.db` and `Level.db`

Please use one of these two options

**Comment by Olshansk** [comment]
tl;dr @adshmh Could you or someone on your team take this on?

@andrewnguyen22 did a detailed job at explaining what we need from a basic keybase solution. This is a precursor that can be worked on in parallel with the CLI being built by @deblasis: https://github.com/pokt-network/pocket/pull/169/files

The keybase doesn't need to be anything fancy, and I think it can actually be _copy-pastad_ or extracted from V0, tendermint or ignite CLI? For example, just using [main/client/keys](https://github.com/cosmos/cosmos-sdk/tree/main/client/keys) and taking out the pieces we need.

**Comment by Olshansk** [comment]
@Jasonyou1995 When you have time to tackle things, one thing I would suggest is doing a bit of research & learning about what a keybase is and how it works. For example, see the summary of the research @deblasis did on the state of RPC in the crypto industry here: https://github.com/pokt-network/pocket/discussions/213

As someone who has never built one myself, here is how I would go about it:
1. Get some high level context googling / researching and reading through, for example, this: https://keybase.io/docs/api/1.0
2. Seeing what their API / specs are like: https://keybase.io/docs/api/1.0/call/key/add
3. Installing the Pocket V0 CLI and seeing the tools available there as a reference
![Screen Shot 2022-10-28 at 10 29 01 AM](https://user-images.githubusercontent.com/1892194/198697215-a5d895c1-2a52-4378-84be-74414028c68a.png)
4. I would then look at the "latest and greatest" tools in the market when it comes to node interaction (e.g. https://ignite.com/)
5. After diving into the code, I found this: https://github.com/ignite/cli/blob/7996b0a3ea0ea528e9558542defcd47c4cd6f1b5/ignite/pkg/cosmoscmd/root.go#L246
![Screen Shot 2022-10-28 at 10 31 13 AM](https://user-images.githubusercontent.com/1892194/198697565-73abb5a8-0266-467d-9180-b44b2fcdf592.png)
6. It looks like (based on the import), that they use cosmos, which led me here: https://github.com/cosmos/cosmos-sdk/tree/main/client/keys
7. Since the keybase isn't something very unique to Pocket, but something we 100% need, and because v0 is a fork of cosmos, I think that using this as a reference is the way to go.
8. I would personally copy-pasta only the parts we need (e.g. create in this PR), make w/e improvements I find, and customize it to Pocket's utility.

Lmk if that's enough context to get you going.

**Comment by Jasonyou1995** [comment]
I'm working on this issue now! Thank you so much for the detailed guide. It's very helpful for me to get started.

**Comment by Olshansk** [comment]
@Jasonyou1995 I've removed `entrypoint` from this PR. 

Check out the demo @deblasis documented in https://github.com/pokt-network/pocket/pull/169

--- 

In v0, here are the entrypoints for `create` and `delete`

<img width="1608" alt="Screen Shot 2022-11-03 at 10 09 40 AM" src="https://user-images.githubusercontent.com/1892194/199788704-a638164e-8afd-4fa2-a112-3052ca0df9a3.png">

--- 

In V1, @deblasis set us up for something equally as simple:

![199353588-46f3987f-49e0-445d-b229-39682d4af881](https://user-images.githubusercontent.com/1892194/199788969-295af477-caa7-42db-acc9-7346b2676478.png)

Also, now that we merged in #177, you can search for `TODO(#150)` in the codebase and find places that need to be updated. https://github.com/pokt-network/pocket/search?q=%22TODO%28%23150%29%22


<img width="755" alt="Screen Shot 2022-11-03 at 10 12 20 AM" src="https://user-images.githubusercontent.com/1892194/199789267-542687e1-f206-4183-968b-dbb602f37155.png">

**Comment by Jasonyou1995** [comment]
Thanks for the clarification @Olshansk! Here are some questions I have:
1. I assume I'm implementing the miniature Keybase in the utility package to update the #150 marked functions with Keybase functions (create and delete).
2. My local Client is V0, and I couldn't see how the V1 entry point change will change how I implement Keybase. Could you please explain a little? Thanks!

**Comment by Olshansk** [comment]
1. It'll be part of the + some local (host) state you need to maintain. The utility module will simply get the private key.
2. It will not, just a different codebase but similar idea.

**Comment by Jasonyou1995** [comment]
Notes

* Discuss with _Alessandro_ integration and general requirements.
* Security requirements: documenting security tasks for future issues.
* Current stage: V0 entry point integration is expected. (potentially V1 integration)

**Comment by Olshansk** [comment]
> * Discuss with _Alessandro_ for integration and general requirements.

Per my comment above, I also believe the demo @deblasis put together is a great starting point too: https://github.com/pokt-network/pocket/issues/150#issuecomment-1302425891

> * Security requirements: documenting security tasks for future issues.

This is VERY optional if you find / learn something interesting along the way.

For example, see the discussion related to `ASCII armored JSON files` here: https://docs.pokt.network/node/setup/

> * Current stage: V0 entry point integration is expected. (potentially V1 integration)

We only expect an integration with V1 (i.e. this repo). The entrypoint is the CLI @deblasis is working on. The approach will be similar (though not identical) to that of V0.

**Comment by Olshansk** [comment]
Came by this while scowering other repos: https://github.com/celestiaorg/celestia-node/blob/main/libs/keystore/keystore.go

Could be an additional good reference for our keybase/keystore

**Comment by Olshansk** [comment]
See #338 as a reference for future iterations of this work.
